src/toast/tod/sss.py

```diff
@@ -0,0 +1,242 @@
+            rank = 0
+            if comm is not None:
+                rank = comm.rank
+            site = self._get_from_obs("site_id", obs)
+            # weather = self._get_from_obs("weather", obs)
+
+            # Get the observation time span and initialize the weather
+            # object if one is provided.
+            times = tod.local_times()
+            tmin = times[0]
+            tmax = times[-1]
+            tmin_tot = tmin
+            tmax_tot = tmax
+            if comm is not None:
+                tmin_tot = comm.allreduce(tmin, op=MPI.MIN)
+                tmax_tot = comm.allreduce(tmax, op=MPI.MAX)
+            # weather.set(site, self._realization, tmin_tot)
+
+            key1, key2, counter1, counter2 = self._get_rng_keys(obs)
+
+            if comm is not None:
+                comm.Barrier()
+            if rank == 0:
+                log.info("{}Setting up SSS simulation".format(prefix))
+
+            tmr = Timer()
+            if self._report_timing:
+                if comm is not None:
+                    comm.Barrier()
+                tmr.start()
+
+            sssmap = self._simulate_sss(key1, key2, counter1, counter2)
+
+            self._observe_sss(sssmap, tod, comm, prefix)
+
+            del sssmap
+
+        if self._report_timing:
+            if comm is not None:
+                comm.Barrier()
+            if rank == 0:
+                tmr.stop()
+                tmr.report(
+                    "{}Simulated and observed scan-synchronous signal" "".format(prefix)
+                )
+        return
+
+    def _get_from_obs(self, name, obs):
+        """ Extract value for name from observation.
+
+        If name is not defined in observation, raise an exception.
+
+        """
+        if name not in obs:
+            raise RuntimeError(
+                "Error simulating SSS: observation " 'does not define "{}"'.format(name)
+            )
+        return obs[name]
+
+    def _get_rng_keys(self, obs):
+        """
+        The random number generator accepts a key and a counter,
+        each made of two 64bit integers.
+        Following tod_math.py we set
+        key1 = realization * 2^32 + telescope * 2^16 + component
+        key2 = obsindx * 2^32
+        counter1 = hierarchical cone counter
+        counter2 = sample in stream (incremented internally in the atm code)
+        """
+        telescope = self._get_from_obs("telescope_id", obs)
+        site = self._get_from_obs("site_id", obs)
+        obsindx = self._get_from_obs("id", obs)
+        key1 = self._realization * 2 ** 32 + telescope * 2 ** 16 + self._component
+        key2 = site * 2 ** 16 + obsindx
+        counter1 = 0
+        counter2 = 0
+        return key1, key2, counter1, counter2
+
+    def _simulate_sss(self, key1, key2, counter1, counter2):
+        """
+        Create a map of the ground signal to observe with all detectors
+        """
+        npix = 12 * self._nside ** 2
+        sssmap = random(
```

Comment:
As long as the resolution of the ground signal map is always low this is fine.  If every process was allocating an NSIDE 4096 map this would be a problem.  Is the ground signal always smoothly varying?  Presumably this is a feature of the environment, independent of the telescope aperature and scan speed...